Mathlib/Tactic/ReduceModChar.lean

```diff
@@ -44,26 +45,62 @@ open Mathlib.Meta.NormNum
+  let ⟨ze, ne, pe⟩ ← Result.toInt instRing (← Mathlib.Meta.NormNum.derive e)
+  let rr ← evalIntMod.go _ _ ze q(IsInt.raw_refl $ne) _ <|
+    .isNat q(instAddMonoidWithOne) _ q(isNat_natCast _ _ (IsNat.raw_refl $n'))
+  let ⟨zr, nr, pr⟩ ← rr.toInt q(Int.instRing)
+  return .isInt instRing nr zr q(CharP.isInt_of_mod $instCharP $pe $pn $pr)
+
+mutual
+
+  /-- Given an expression of the form `a ^ b` in a ring of characteristic `n`, reduces `a`
+      modulo `n` recursively and then calculates `a ^ b` using fast modular exponentiation. -/
+  partial def normPow {α : Q(Type u)} (n n' : Q(ℕ)) (pn : Q(IsNat «$n» «$n'»)) (e : Q($α))
+      (instRing : Q(Ring $α)) (instCharP : Q(CharP $α $n)) : MetaM (Result e) := do
+    let .app (.app (f : Q($α → ℕ → $α)) (a : Q($α))) (b : Q(ℕ)) ← whnfR e | failure
+    let .isNat sα na pa ← normIntNumeral' n n' pn a instRing instCharP | failure
+    let ⟨nb, pb⟩ ← Mathlib.Meta.NormNum.deriveNat b q(instAddMonoidWithOneNat)
+    guard <|← withDefault <| withNewMCtxDepth <| isDefEq f q(HPow.hPow (α := $α))
+    haveI' : $e =Q $a ^ $b := ⟨⟩
+    haveI' : $f =Q HPow.hPow := ⟨⟩
+    have ⟨c, r⟩ := evalNatPowMod na nb n'
+    assumeInstancesCommute
+    return .isNat sα c q(CharP.isNat_pow (f := $f) $instCharP (.refl $f) $pa $pb $pn $r)
+
+  /-- If `e` is of the form `a ^ b`, reduce it using fast modular exponentation, otherwise
+      reduce it using `norm_num`. -/
+  partial def normIntNumeral' {α : Q(Type u)} (n n' : Q(ℕ)) (pn : Q(IsNat «$n» «$n'»))
+      (e : Q($α)) (instRing : Q(Ring $α)) (instCharP : Q(CharP $α $n)) : MetaM (Result e) :=
+    normPow n n' pn e instRing instCharP <|> normBareNumeral n n' pn e instRing instCharP
+
+end
+
 lemma CharP.intCast_eq_mod (R : Type _) [Ring R] (p : ℕ) [CharP R p] (k : ℤ) :
     (k : R) = (k % p : ℤ) := by
   calc
     (k : R) = ↑(k % p + p * (k / p)) := by rw [Int.emod_add_ediv]
     _ = ↑(k % p) := by simp [CharP.cast_eq_zero R]
 
-lemma CharP.isInt_of_mod {e' r : ℤ} {α : Type _} [Ring α] {n n' : ℕ} (inst : CharP α n) {e : α}
-    (he : IsInt e e') (hn : IsNat n n') (h₂ : IsInt (e' % n') r) : IsInt e r :=
-  ⟨by rw [he.out, CharP.intCast_eq_mod α n, show n = n' from hn.out, h₂.out, Int.cast_id]⟩
-
 /-- Given an integral expression `e : t` such that `t` is a ring of characteristic `n`,
 reduce `e` modulo `n`. -/
 partial def normIntNumeral {α : Q(Type u)} (n : Q(ℕ)) (e : Q($α)) (instRing : Q(Ring $α))
     (instCharP : Q(CharP $α $n)) : MetaM (Result e) := do
-  let ⟨ze, ne, pe⟩ ← Result.toInt instRing (← Mathlib.Meta.NormNum.derive e)
   let ⟨n', pn⟩ ← deriveNat n q(instAddMonoidWithOneNat)
-  let rr ← evalIntMod.go _ _ ze q(IsInt.raw_refl $ne) _ <|
-    .isNat q(instAddMonoidWithOne) _ q(isNat_natCast _ _ (IsNat.raw_refl $n'))
-  let ⟨zr, nr, pr⟩ ← rr.toInt q(Int.instRing)
-  return .isInt instRing nr zr q(CharP.isInt_of_mod $instCharP $pe $pn $pr)
+  normIntNumeral' n n' pn e instRing instCharP
```

Comment:
```suggestion
  normIntNumeral' n n' pn e _ instCharP
```
and make `(_ : Q(Ring $α)` the parameter. This also doesn't need to be `partial` though maybe it (and above) should remain so for psychological reasons.